boards/mulle/board.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (C) 2014 Eistec AB
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     board_k60
+ * @{
+ *
+ * @file
+ * @brief       Board specific implementations for the Mulle board
+ *
+ * @author      Joakim Gebart <joakim.gebart@eistec.se>
+ *
+ * @}
+ */
+
+#include <stddef.h> /* for NULL */
+#include <stdio.h>
+#include "board.h"
+#include "cpu.h"
+#include "mcg.h"
+#include "periph/gpio.h"
+#include "periph/uart.h"
+#include "periph/rtc.h"
+#include "devicemap.h"
+
+/**
+ * @brief Initialize the boards on-board LEDs
+ *
+ * The LEDs are initialized here in order to be able to use them in the early
+ * boot for diagnostics.
+ *
+ */
+static inline void leds_init(void);
+
+/** @brief Initialize the GPIO pins controlling the power switches. */
+static inline void power_pins_init(void);
+
+/**
+ * @brief Set clock prescalers to safe values
+ *
+ * This should be done before switching to FLL/PLL as clock source to ensure
+ * that all clocks remain within the specified limits.
+ */
+static inline void set_safe_clock_dividers(void);
+
+/** @brief Set the FLL source clock to RTC32k */
+static inline void set_fll_source(void);
+
+
+void board_init(void)
+{
+    /* initialize the boards LEDs, this is done first for debugging purposes */
+    leds_init();
+
+    LED_RED_ON;
+
+    /* Initialize RTC oscillator as early as possible since we are using it as a
+     * base clock for the FLL.
+     * It takes a while to stabilize the oscillator, therefore we do this as
+     * soon as possible during boot in order to let it stabilize while other
+     * stuff is initializing. */
+    /* If the clock is not stable then the UART will have the wrong baud rate
+     * for debug prints as well */
+    rtc_init();
+
+    /* Set up clocks */
+    set_safe_clock_dividers();
+
+    set_fll_source();
+
+    kinetis_mcg_set_mode(KINETIS_MCG_FEE);
+
+    /* At this point we need to wait for 1 ms until the clock is stable.
+     * Since the clock is not yet stable we can only guess how long we must
+     * wait. I have tried to make this as short as possible but still being able
+     * to read the initialization messages written on the UART.
+     * (If the clock is not stable all UART output is garbled until it has
+     * stabilized) */
+    for (int i = 0; i < 100000; ++i) {
+        asm volatile("nop\n");
+    }
```

Comment:
I am not sure if this may affect other communications as well, such as SPI. Too low clock frequency is rarely a problem for SPI and I2C, but I do not know the exact behaviour of the unstable clock and whether it may cause too short delays in some peripheral drivers.